File: Sean的leetcode/Leetcode/Sliding Window Maximum.cpp
//https://leetcode.com/discuss/46578/java-o-n-solution-using-deque-with-explanation
//Time complexity is O(n) because each element is executed at most one time push_back() and pop_front()/pop_back(), which is O(2n) = O(n)
//Space:O(n)
//deque = double-ended queue: can insert/remove element from both sides
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int>res;
        int n = nums.size();
        if (!n || !k) return res;
        
        deque<int>window;

        for (int i = 0; i < n; i++)
        {
            if (!window.empty() && window.front() < i - k + 1)
                window.pop_front();

            while (!window.empty() && nums[window.back()] < nums[i])
                window.pop_back();

            window.push_back(i);
            if (i >= k-1)
                res.push_back(nums[window.front()]);
        }
        return res;
    }
};

File: MyLeetCode/python/Sliding Window Maximum.py
###
# Deque
# Time Complexity: O(n) every element insert/delete once
# Space Complexity: O(k)
###
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        from collections import deque
        w = deque()
        re = []
        for i, n in enumerate(nums):
            # move left end
            if w and w[0] == nums[i - k]:
                w.popleft()
            # insert element
            while w and w[-1] < n:
                w.pop()
            w.append(n)
            if i >= k-1:
                re.append(w[0])
        return re
        
        